utility/utility.go

```diff
@@ -22,3 +31,168 @@ func ProfileExportFileNameNoPath(info profileutil.ProvisioningProfileInfoModel)
+	return xcodeProj.TargetBuildSettings(target, configuration, customOptions...)
+}
+
+// BuildableTargetPlatform ...
+func BuildableTargetPlatform(
+	xcodeProj *xcodeproj.XcodeProj,
+	scheme *xcscheme.Scheme,
+	configurationName string,
+	provider TargetBuildSettingsProvider,
+) (Platform, error) {
+	archiveEntry, ok := scheme.AppBuildActionEntry()
+	if !ok {
+		return "", fmt.Errorf("archivable entry not found in project: %s, scheme: %s", xcodeProj.Path, scheme.Name)
+	}
+
+	mainTarget, ok := xcodeProj.Proj.Target(archiveEntry.BuildableReference.BlueprintIdentifier)
+	if !ok {
+		return "", fmt.Errorf("target not found: %s", archiveEntry.BuildableReference.BlueprintIdentifier)
+	}
+
+	settings, err := provider.TargetBuildSettings(xcodeProj, mainTarget.Name, configurationName)
+	if err != nil {
+		return "", fmt.Errorf("failed to get target (%s) build settings: %s", mainTarget.Name, err)
+	}
+
+	return getPlatform(settings)
+}
+
+func getPlatform(buildSettings serialized.Object) (Platform, error) {
+	/*
+		Xcode help:
+		Base SDK (SDKROOT)
+		The name or path of the base SDK being used during the build.
+		The product will be built against the headers and libraries located inside the indicated SDK.
+		This path will be prepended to all search paths, and will be passed through the environment to the compiler and linker.
+		Additional SDKs can be specified in the Additional SDKs (ADDITIONAL_SDKS) setting.
+
+		Examples:
+		- /Applications/Xcode.app/Contents/Developer/Platforms/AppleTVOS.platform/Developer/SDKs/AppleTVOS.sdk
+		- /Applications/Xcode.app/Contents/Developer/Platforms/AppleTVSimulator.platform/Developer/SDKs/AppleTVSimulator13.4.sdk
+		- /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS13.4.sdk
+		- /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator.sdk
+		- /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk
+		- /Applications/Xcode.app/Contents/Developer/Platforms/WatchOS.platform/Developer/SDKs/WatchOS.sdk
+		- /Applications/Xcode.app/Contents/Developer/Platforms/WatchSimulator.platform/Developer/SDKs/WatchSimulator.sdk
+		- iphoneos
+		- macosx
+		- appletvos
+		- watchos
+	*/
+	sdk, err := buildSettings.String("SDKROOT")
+	if err != nil {
+		return "", fmt.Errorf("failed to get SDKROOT: %s", err)
+	}
+
+	sdk = strings.ToLower(sdk)
+	if filepath.Ext(sdk) == ".sdk" {
+		sdk = filepath.Base(sdk)
+	}
+
+	switch {
+	case strings.HasPrefix(sdk, "iphoneos"):
+		return iOS, nil
+	case strings.HasPrefix(sdk, "macosx"):
+		return osX, nil
+	case strings.HasPrefix(sdk, "appletvos"):
+		return tvOS, nil
+	case strings.HasPrefix(sdk, "watchos"):
+		return watchOS, nil
+	default:
+		return "", fmt.Errorf("unkown SDKROOT: %s", sdk)
+	}
+}
+
+// OpenArchivableProject ...
+func OpenArchivableProject(pth, schemeName, configurationName string) (*xcodeproj.XcodeProj, *xcscheme.Scheme, string, error) {
```

Comment:
you do not need a separate `OpenArchivableProject` and `OpenArchivableWorkspace`.

`schemeint.Scheme` accepts both a workspace and project path, whatever is provided,  it figures out the correct root project (which is connected to the scheme).

The issue with the current implementation is at the `OpenArchivableWorkspace`: In the case of a workspace project, a given scheme can be located both under the .xcworkspace or under one of the included .xcproject files, this is handled correctly by the `schemeint.Scheme`, but `OpenArchivableWorkspace` is not aware of the schemes under the `.xcworkspace`.

Reply:
![Screenshot 2022-01-28 at 10 10 57](https://user-images.githubusercontent.com/12427060/151519168-bde8b2ca-7adb-4a84-b635-8d210514b993.png)